File: python_backend/analytics/services/grid_service.py
from django.utils import timezone
from dateutil import parser
from analytics.models import Match, Team, Player, PlayerStats, TeamStats, Draft, ChampionPool
from .grid_api_client import GridAPIClient

class GridService:
    def __init__(self):
        self.client = GridAPIClient()

    def ingest_match(self, grid_match_id: str):
        """
        Ingest a match from GRID using official APIs.
        """
        print(f"Ingesting match from GRID: {grid_match_id}...")
        
        # 1. Fetch data from GRID APIs
        series_response = self.client.get_series_context(grid_match_id)
        match_stats = self.client.get_match_stats(grid_match_id)
        
        # Extract series context from GraphQL response
        series_context = series_response.get('data', {}).get('series')
        
        if not series_context or not match_stats:
            print(f"Failed to fetch data for match {grid_match_id}")
            return None

        # 2. Get or Create Teams
        team_map = {} # GRID ID -> Team Object
        for team_entry in series_context['teams']:
            team_data = team_entry['team']
            team, _ = Team.objects.get_or_create(
                name=team_data['name'],
                defaults={
                    'region': 'Unknown', # GraphQL snippet didn't include region
                    'league': series_context['tournament']['name']
                }
            )
            team_map[team_data['id']] = team

        # 3. Create Match
        winner_id = match_stats.get('winner') # Winner usually comes from stats/result
        winner = team_map.get(winner_id)
        
        match, created = Match.objects.get_or_create(
            grid_match_id=grid_match_id,
            defaults={
                "date": parser.parse(series_context['startTimeScheduled']),
                "patch": match_stats.get('patch', 'Unknown'),
                "duration": match_stats.get('duration', 0),
                "winner": winner,
                "format_type": series_context['format']['type'],
                "tournament_name": series_context['tournament']['name'],
                "game_title": series_context['title']['nameShortened']
            }
        )
        
        if not created:
            print(f"Match {grid_match_id} already exists. Skipping.")
            return match

        # 4. Create Players & Stats
        # Note: GraphQL series context has limited player info, using stats feed for players
        for p_stats in match_stats['playerStats']:
            # Find which team this player belongs to from stats feed if available,
            # or try to map based on match_stats teams
            t_stats = next((ts for ts in match_stats['teamStats'] if any(ps['playerId'] == p_stats['playerId'] for ps in match_stats['playerStats'])), None)
            
            # Simple mapping for demo/implementation:
            # In a real scenario, GRID provides detailed player/team mapping in the series context.
            # Here we assume player ID can be used to find or create.
            
            # For this implementation, we'll try to get the team from the stats feed team data
            team_id = next((ts['teamId'] for ts in match_stats['teamStats'] if ts['teamId'] in team_map), None)
            # This is a bit simplified, in reality match_stats would link player to team
            
            player, _ = Player.objects.get_or_create(
                identifier=p_stats.get('playerName', f"Player-{p_stats['playerId']}"),
                defaults={"role": "Unknown", "team": team_map.get(list(team_map.keys())[0])} # Fallback
            )
            
            PlayerStats.objects.create(
                match=match,
                player=player,
                kills=p_stats['kills'],
                deaths=p_stats['deaths'],
                assists=p_stats['assists'],
                cs=p_stats['cs'],
                gold_earned=p_stats['gold'],
                positioning_score=0.85
            )

        # 5. Create Team Stats
        for t_stats in match_stats['teamStats']:
            team = team_map.get(t_stats['teamId'])
            if team:
                TeamStats.objects.create(
                    match=match,
                    team=team,
                    barons=t_stats['barons'],
                    dragons=t_stats['dragons'],
                    towers=t_stats['towers'],
                    gold_diff_15=t_stats['goldDiffAt15']
                )

        # 6. Create Drafts
        for team_id, draft_data in match_stats['draft'].items():
            team = team_map.get(team_id)
            if team:
                Draft.objects.create(
                    match=match,
                    team=team,
                    bans=draft_data['bans'],
                    picks=draft_data['picks'],
                    win_probability=0.5 # Default, updated by AnalyticsService
                )
            
        print(f"Successfully ingested match {match.grid_match_id}")
        return match

File: src/services/grid.service.ts
import { prisma } from '../lib/prisma';
import { GridAPIClient } from './grid-api-client';

export class GridService {
  private client: GridAPIClient;

  constructor() {
    this.client = new GridAPIClient();
  }

  async ingestMatch(gridMatchId: string) {
    console.log(`Ingesting match from GRID: ${gridMatchId}...`);
    const seriesResponse = await this.client.getSeriesContext(gridMatchId);
    const matchStats = await this.client.getMatchStats(gridMatchId);

    const seriesContext = seriesResponse?.data?.series;

    if (!seriesContext || !matchStats) {
      console.error(`Failed to fetch data for match ${gridMatchId}`);
      return null;
    }

    // 1b. Create or Update Series
    const series = await prisma.series.upsert({
      where: { gridSeriesId: seriesContext.id },
      update: {
        tournamentName: seriesContext.tournament.name,
        format: seriesContext.format.type,
        startTime: new Date(seriesContext.startTimeScheduled),
      },
      create: {
        gridSeriesId: seriesContext.id,
        tournamentName: seriesContext.tournament.name,
        format: seriesContext.format.type,
        startTime: new Date(seriesContext.startTimeScheduled),
      },
    });

    // 2. Get or Create Teams
    const teamMap: Record<string, any> = {}; // GRID ID -> Team Object
    for (const teamEntry of seriesContext.teams) {
      const teamData = teamEntry.team;
      const team = await prisma.team.upsert({
        where: { name: teamData.name },
        update: {
          region: 'Unknown',
          league: seriesContext.tournament.name,
          series: {
            connect: { id: series.id }
          }
        },
        create: {
          name: teamData.name,
          region: 'Unknown',
          league: seriesContext.tournament.name,
          series: {
            connect: { id: series.id }
          }
        },
      });
      teamMap[teamData.id] = team;
    }

    // 3. Create Match
    const winnerId = matchStats.winner;
    const winner = teamMap[winnerId];

    const match = await prisma.match.upsert({
      where: { gridMatchId },
      update: {
        seriesId: series.id,
      },
      create: {
        gridMatchId,
        seriesId: series.id,
        date: new Date(seriesContext.startTimeScheduled),
        patch: matchStats.patch || 'Unknown',
        duration: matchStats.duration || 0,
        winnerId: winner?.id,
        formatType: seriesContext.format.type,
        tournamentName: seriesContext.tournament.name,
        gameTitle: seriesContext.title.nameShortened,
      },
    });

    // Check if player stats already exist (simplified check)
    const existingStatsCount = await prisma.playerStats.count({
      where: { matchId: match.id },
    });

    if (existingStatsCount > 0) {
      console.log(`Match ${gridMatchId} already has stats. Skipping ingestion of stats.`);
      return match;
    }

    // 4. Create Players & Stats
    // Assuming team index 0 is first team, team index 1 is second team for mock mapping
    const teamIds = Object.keys(teamMap);
    
    for (let i = 0; i < matchStats.playerStats.length; i++) {
      const pStats = matchStats.playerStats[i];
      // In a real scenario, we'd map players correctly. Here we distribute them.
      const assignedTeamId = i < 5 ? teamIds[0] : teamIds[1];
      const team = teamMap[assignedTeamId];

      const player = await prisma.player.upsert({
        where: {
          teamId_identifier: {
            teamId: team.id,
            identifier: pStats.playerName || `Player-${pStats.playerId}`,
          },
        },
        update: {},
        create: {
          identifier: pStats.playerName || `Player-${pStats.playerId}`,
          role: 'Unknown',
          teamId: team.id,
        },
      });

      await prisma.playerStats.create({
        data: {
          matchId: match.id,
          playerId: player.id,
          kills: pStats.kills,
          deaths: pStats.deaths,
          assists: pStats.assists,
          cs: pStats.cs,
          goldEarned: pStats.gold,
          positioningScore: 0.85,
        },
      });
    }

    // 5. Create Team Stats
    for (const tStats of matchStats.teamStats) {
      const team = teamMap[tStats.teamId];
      if (team) {
        await prisma.teamStats.create({
          data: {
            matchId: match.id,
            teamId: team.id,
            barons: tStats.barons,
            dragons: tStats.dragons,
            towers: tStats.towers,
            goldDiff15: tStats.goldDiffAt15,
          },
        });
      }
    }

    // 6. Create Drafts
    for (const [teamGridId, draftData] of Object.entries(matchStats.draft)) {
      const team = teamMap[teamGridId];
      if (team) {
        await prisma.draft.create({
          data: {
            matchId: match.id,
            teamId: team.id,
            bans: JSON.stringify(draftData.bans),
            picks: JSON.stringify(draftData.picks),
            winProbability: 0.5,
          },
        });
      }
    }

    console.log(`Successfully ingested match ${match.gridMatchId}`);
    return match;
  }

  /**
   * Get team statistics for scouting report
   */
  async getTeamStatsForScouting(teamName: string) {
    try {
      // Get team from database
      const team = await prisma.team.findFirst({
        where: { name: { contains: teamName } },
        include: {
          _count: {
            select: { players: true }
          }
        }
      });

      if (!team) {
        return null;
      }

      // Get recent matches (via joins)
      const recentMatches = await prisma.match.findMany({
        where: {
          OR: [
            { winnerId: team.id },
            {
              teamStats: {
                some: { teamId: team.id }
              }
            }
          ]
        },
        include: {
          series: {
            include: {
              teams: true
            }
          },
          winner: true,
          teamStats: {
            where: { teamId: team.id }
          },
          drafts: {
            where: { teamId: team.id }
          }
        },
        orderBy: { date: 'desc' },
        take: 20
      });

      // Calculate win/loss
      const wins = recentMatches.filter(m => m.winnerId === team.id).length;
      const losses = recentMatches.length - wins;
      const winRate = recentMatches.length > 0 ? (wins / recentMatches.length) * 100 : 0;

      // Extract champion pool from drafts
      const championPicks = new Map<string, number>();
      const championBans = new Map<string, number>();

      for (const match of recentMatches) {
        const draft = match.drafts.find(d => d.teamId === team.id);
        if (draft) {
          try {
            const picks = JSON.parse(draft.picks as string);
            const bans = JSON.parse(draft.bans as string);
            
            picks.forEach((champ: string) => {
              championPicks.set(champ, (championPicks.get(champ) || 0) + 1);
            });
            
            bans.forEach((champ: string) => {
              championBans.set(champ, (championBans.get(champ) || 0) + 1);
            });
          } catch (e) {
            // Skip malformed draft data
          }
        }
      }

      // Get top champions
      const topPicks = Array.from(championPicks.entries())
        .sort((a, b) => b[1] - a[1])
        .slice(0, 10)
        .map(([name, count]) => ({ name, count }));

      const topBans = Array.from(championBans.entries())
        .sort((a, b) => b[1] - a[1])
        .slice(0, 5)
        .map(([name, count]) => ({ name, count }));

      // Calculate average game time
      const avgGameTime = recentMatches.length > 0
        ? recentMatches.reduce((sum, m) => sum + (m.duration || 0), 0) / recentMatches.length
        : 0;

      // Determine trend (simple: last 5 vs previous 5)
      const recent5 = recentMatches.slice(0, 5);
      const previous5 = recentMatches.slice(5, 10);
      const recent5Wins = recent5.filter(m => m.winnerId === team.id).length;
      const previous5Wins = previous5.filter(m => m.winnerId === team.id).length;
      
      let trend: 'improving' | 'declining' | 'stable' = 'stable';
      if (recent5Wins > previous5Wins + 1) trend = 'improving';
      else if (recent5Wins < previous5Wins - 1) trend = 'declining';

      return {
        teamId: team.id,
        teamName: team.name,
        region: team.region,
        league: team.league,
        wins,
        losses,
        winRate,
        totalMatches: recentMatches.length,
        topPicks,
        topBans,
        avgGameTime: Math.round(avgGameTime),
        trend,
        recentMatches: recentMatches.slice(0, 5).map(m => ({
          date: m.date,
          opponent: m.series.teams.find(t => t.id !== team.id)?.name || 'Unknown',
          result: m.winnerId === team.id ? 'W' : 'L',
          duration: m.duration,
        }))
      };
    } catch (error) {
      console.error('Error fetching team stats for scouting:', error);
      return null;
    }
  }
}

